### ISSUE TITLE
DDP/MP not yielding nontrivial speedup

### ISSUE THREAD

**Issue by timtody** [pytorch/pytorch#31360]
## 🐛 Bug

Following the tutorial from https://pytorch.org/tutorials/intermediate/ddp_tutorial.html I implemented a distributed policy gradient reinforcement learning algorithm.  Using the script below I benchmarked 1000 steps on a simple gym environment and recorded the time per worker. Since I'm on a six core machine I was expecting a nontrivial speedup per global step in the order of 1 <= num_processes <= 6. This is not the case (see output below). 

I'm not trying to do any learning here, just benchmarking. I observed the same behavior when using DDP and even asynchonous code from the hogwild example at https://pytorch.org/docs/stable/notes/multiprocessing.html .
## To Reproduce

1. Run script below with different values for num_processes 

```
import sys
import os
import tempfile
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim
import torch.multiprocessing as mp
from torch.nn.parallel import DistributedDataParallel as DDP
from models import DNNPolicy, FCPolicy
import time
import gym


def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'

    dist.init_process_group("gloo", rank=rank, world_size=world_size)
    torch.manual_seed(42)

def cleanup():
    dist.destroy_process_group()

def main(rank, world_size):
    setup(rank, world_size)
    env = gym.make('CartPole-v0')
    policy = FCPolicy(env)
    state = env.reset()
    loop_start = time.time()
    for i in range(1000):
        policy(torch.randn(4))
        env.render()
    loop_end = time.time()
    print("loop time worker:", rank, "nprocs:", world_size, loop_end - loop_start, "s")

def run_demo(demo_fn, world_size):
    mp.spawn(demo_fn,
             args=(world_size,),
             nprocs=world_size,
             join=True)

if __name__ == "__main__":
    for n_procs in range(6):
        run_demo(main, n_procs+1)
```
2. Results in:

loop time worker: 0 nprocs: 1 5.78078556060791 s
loop time worker: 0 nprocs: 2 6.326328992843628 s
loop time worker: 1 nprocs: 2 6.847925186157227 s
loop time worker: 2 nprocs: 3 8.730802536010742 s
loop time worker: 1 nprocs: 3 9.05618143081665 s
loop time worker: 0 nprocs: 3 9.310847759246826 s
loop time worker: 3 nprocs: 4 11.378642320632935 s
loop time worker: 0 nprocs: 4 11.415692329406738 s
loop time worker: 1 nprocs: 4 11.841736316680908 s
loop time worker: 2 nprocs: 4 12.169833421707153 s
loop time worker: 3 nprocs: 5 17.826525449752808 s
loop time worker: 2 nprocs: 5 18.16375470161438 s
loop time worker: 0 nprocs: 5 18.549652814865112 s
loop time worker: 1 nprocs: 5 18.617838859558105 s
loop time worker: 4 nprocs: 5 18.68006205558777 s
loop time worker: 5 nprocs: 6 17.35166072845459 s
loop time worker: 4 nprocs: 6 18.644100189208984 s
loop time worker: 0 nprocs: 6 18.832598447799683 s
loop time worker: 2 nprocs: 6 19.0396511554718 s
loop time worker: 3 nprocs: 6 19.1573805809021 s
loop time worker: 1 nprocs: 6 19.286489248275757 s
## Expected behavior

Loop time staying roughly constant for increasing amount of workers.
## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.17.4
[pip] torch==1.3.1
[conda] torch                     1.3.1                    pypi_0    pypi

**Comment by mruberry** [comment]
@timtody It's hard to diagnose every model, but it's not clear you should expect a significant speedup in this case. 

Can you look at CPU utilization when you're running with a single process vs. when you're running with multiple processes? Maybe the single process is using multiple cores.
